### PR TITLE
Next LTO attempt

### PR DESCRIPTION
**Pull request by vanhauser-thc**
Don't merge yet, I first want to be sure it compiles and see which test cases work and don't and if I can fix that. thanks!

### CODE REVIEW

**Comment by jonathanmetzman** [comment]
Could you pull from master please? I've made some changes to CI and I want to see if they're working. And you might even find the changes useful!

**Comment by vanhauser-thc** [comment]
> Could you pull from master please? I've made some changes to CI and I want to see if they're working. And you might even find the changes useful!

did you push them? because in the commit history there is nothing from you and when I try to pull I get "There isn’t anything to compare. AFLplusplus:master is up to date with all commits from google:master. "

**Comment by vanhauser-thc** [comment]
What can be done about these errors?

```
 Error response from daemon: manifest for gcr.io/fuzzbench/builders/aflplusplus_lto_mopt/sqlite3_ossfuzz-intermediate:latest not found: manifest unknown: Failed to fetch "latest" from request "/v2/fuzzbench/builders/aflplusplus_lto_mopt/sqlite3_ossfuzz-intermediate/manifests/latest".
make: *** [.pull-aflplusplus_lto_mopt-sqlite3_ossfuzz-oss-fuzz-builder-intermediate] Error 1
docker/build.mk:303: recipe for target '.pull-aflplusplus_lto_mopt-sqlite3_ossfuzz-oss-fuzz-builder-intermediate' failed
make: *** Waiting for unfinished jobs....
```

**Comment by vanhauser-thc** [comment]
For the LTO variants everything builds with the exception of:

 - curl_curl_fuzzer_http - the compilation process runs out of memory. dunno how much memory would be needed in this setup, but it is > 35GB (see issue #320 )
- tcpdump-4.9.0 - however that does not seem to be an issue of LTO:
```
checking for accept4... yes
checking abstract socket namespace... yes
checking for XML_ParserCreate_MM in -lexpat... no
configure: error: Explicitly requested expat but expat not found
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/src/fuzzer.py", line 23, in build
    aflplusplus_fuzzer.build("lto", "nozero")
  File "/src/fuzzers/aflplusplus/fuzzer.py", line 112, in build
    utils.build_benchmark()
  File "/src/fuzzers/utils.py", line 52, in build_benchmark
    subprocess.check_call(['/bin/bash', '-ex', build_script], env=env)
  File "/usr/local/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/bin/bash', '-ex', 'benchmark/build.sh']' returned non-zero exit status 1.
The command '/bin/sh -c python3 -u -c "from fuzzers import utils; utils.initialize_flags(); import fuzzer; fuzzer.build()"' returned a non-zero code: 1
make: *** [docker/build.mk:181: .aflplusplus_lto_mopt-tcpdump-4.9.0-builder] Error 1
```

so it would be great if this could be merged. I removed two other variants to make space for these, so there are 8 again:
```
          - aflplusplus_lto_mopt
          - aflplusplus_lto_instrim_mopt
          - aflplusplus_ngram3_cmplog_mopt2_17
          - aflplusplus_instrim_mopt
          - aflplusplus_instrim_mopt2
          - aflplusplus_instrim_cmplog_mopt
          - aflplusplus_instrim_laf_mopt2
          - aflplusplus_instrim_laf_mopt2_17
```

If you could start them together and ignore the ones LTO is not building for that would be great!

**Comment by jonathanmetzman** [comment]
> > Could you pull from master please? I've made some changes to CI and I want to see if they're working. And you might even find the changes useful!
> 
> did you push them? because in the commit history there is nothing from you and when I try to pull I get "There isn’t anything to compare. AFLplusplus:master is up to date with all commits from google:master. "

My apologies, you did actually pull them I was just confused about what it would look like.
The change was that the builds for unaffected fuzzers (e.g. eclipser) are noops and return successfully almost immediately.

> What can be done about these errors?
> 
> ```
>  Error response from daemon: manifest for gcr.io/fuzzbench/builders/aflplusplus_lto_mopt/sqlite3_ossfuzz-intermediate:latest not found: manifest unknown: Failed to fetch "latest" from request "/v2/fuzzbench/builders/aflplusplus_lto_mopt/sqlite3_ossfuzz-intermediate/manifests/latest".
> make: *** [.pull-aflplusplus_lto_mopt-sqlite3_ossfuzz-oss-fuzz-builder-intermediate] Error 1
> docker/build.mk:303: recipe for target '.pull-aflplusplus_lto_mopt-sqlite3_ossfuzz-oss-fuzz-builder-intermediate' failed
> make: *** Waiting for unfinished jobs....
> ```

Where did you see this? As far as I know, the pull failing wont cause the build to stop. There's nothing that can be done about this since until we build the fuzzer in an experiment there is nothing to pull (maybe we could avoid pulling for new fuzzers but this seems like overkill if everything WAI).

**Comment by jonathanmetzman** [comment]
> If you could start them together and ignore the ones LTO is not building for that would be great!

Will do. The backlog grew a bit and so i'm hoping to run this by middle of next week.

In the future, let's do these to make the afl++ integrations simpler:
1. Let's remove old variants in a separate PR. This way is much easier for me to review.
2. If there are fuzzers that have small runtime parameters changed, maybe we shouldn't add them all to CI?

**Comment by jonathanmetzman** [comment]
> Will do. The backlog grew a bit and so i'm hoping to run this by middle of next week.

Actually I'm a bit confused. Why didn't this PR remove the old variants? And which fuzzers should I run in the experiment?

**Comment by vanhauser-thc** [comment]
> > Will do. The backlog grew a bit and so i'm hoping to run this by middle of next week.
> 
> Actually I'm a bit confused. Why didn't this PR remove the old variants? And which fuzzers should I run in the experiment?

the last commit e22ef15 in this PR which is labeled "remove 2 fuzzer variants to make space for LTO" and removes two variants.

the list which are interesting to run is in my previous post: https://github.com/google/fuzzbench/pull/319#issuecomment-626354484

**Comment by vanhauser-thc** [comment]
So I basically spent the whole day (12hours) to get LTO working for every target. Just because build-... works, does not mean run-... works, and because I just have a normal laptop with a normal setup this was tedious. very tedious.
But now (at least on my system) I can run and make all targets. Lets hope I didnt fuck something up :)

**Comment by vanhauser-thc** [comment]
I did it! :)

File: fuzzers/aflplusplus_lto_mopt/builder.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#    sed -i 's/https:/http:/g' /etc/apt/sources.list /etc/apt/sources.list.d/* && \
+RUN apt-get update && \
+    apt-get install -y wget libstdc++-5-dev libexpat1-dev && \
+    apt-get install -y apt-utils apt-transport-https ca-certificates && \
+    echo deb http://apt.llvm.org/xenial/ llvm-toolchain-xenial main >> /etc/apt/sources.list && \
+    echo deb http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu xenial main >> /etc/apt/sources.list && \
+    wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add - && \
+    apt-key adv --recv-keys --keyserver keyserver.ubuntu.com 1E9377A2BA9EF27F && \
+    apt-get update && \
+    apt-get install -y clang-11 clang-11-doc clang-11-examples clangd-11 clang-format-11 clang-tidy-11 clang-tools-11 libc++1-11 libc++-11-dev libc++abi1-11 libc++abi-11-dev libclang1-11 libclang-11-dev libclang-common-11-dev libclang-cpp11 libclang-cpp11-dev libfuzzer-11-dev liblld-11 liblld-11-dev liblldb-11 liblldb-11-dev libllvm11 libomp-11-dev libomp-11-doc libomp5-11 lld-11 lldb-11 llvm-11 llvm-11-dev llvm-11-runtime llvm-11-tools && \
```
Comment (jonathanmetzman) [inline fuzzers/aflplusplus_lto_mopt/builder.Dockerfile:28]: In general better to break this up.

**Comment by jonathanmetzman** [comment]
> So I basically spent the whole day (12hours) to get LTO working for every target. Just because build-... works, does not mean run-... works, and because I just have a normal laptop with a normal setup this was tedious. very tedious.

Heroic effort :-) wow.

**Pull request merged**